### PR TITLE
fix(completion_preview.lua): fix for cursor position for multi line completions

### PR DESCRIPTION
**Pull request by Hashiraee**
The main change is at the end of the `CompletionPreview.on_accept_suggestion()`, I calculate the new cursor position based on the number of lines in the `completion_text`.

So I just:
- Split the the completion_text into lines using vim.split()
- Then update the new cursor position using nvim_win_set_cursor()

Setting the cursor position using `nvim_win_set_cursor()` is a more standard approach, especially if users have certain keybindings for specific keys compared to PR #2.

This fixes Issue #2.

### CODE REVIEW

**Comment by danthomps** [comment]
@amirhhashemi You're right. We have a fix in the discord for it. I'm sure he will add it to the PR soon.

**Comment by Hashiraee** [comment]
Yep, we already have a fix, it does the completion but moves it to a weird spot. I will make the PR now.

**Comment by amirhhashemi** [comment]
The new commit fixed it. LGTM.

**Comment by Hashiraee** [comment]
Can be merged now, Issue #2 is resolved @super-jacob!

**Comment by super-jacob** [comment]
This looks good. I'll wait for @victorw-xyz's review before merging.

**Comment by Hashiraee** [comment]
Added a fix for `get_uft8_length()` using neovim's built-in `vim.fn.strlen()`

**Review by victorw-xyz — APPROVED** [review]
This solves issue #2 without having to check for specific versions of Neovim, which is nice, as we won't have to continue updating this logic

**Comment by b0o** [comment]
> Added a fix for `get_uft8_length()` using neovim's built-in `vim.fn.strlen()`

I was running into the 8000-char limit as well and ended up on this PR. I am questioning whether using `vim.fn.strlen()` or the other methods is necessary at all. Lua's `string.len()` / `#` operator seem to return the byte length of strings with multi-byte characters, the same as the original implementation and `vim.fn.strlen()`.

```lua
str = "😊"
print(#str) -- 4
print(str:len()) -- 4
print(vim.fn.strlen(str)) -- 4
print(#{str:byte(1, -1)}) -- 4
```

Why not just use `#`?

**Comment by 3rd** [comment]
> > Added a fix for `get_uft8_length()` using neovim's built-in `vim.fn.strlen()`
> 
> I was running into the 8000-char limit as well and ended up on this PR.
> 
> I question whether using `vim.fn.strlen()` or the other `string.byte()` method is necessary at all. Lua's `string.len()` / `#` operator already return the byte length of strings with multi-byte characters, the same as the original `string.byte()` implementation and `vim.fn.strlen()`:
> 
> ```lua
> str = "😊"
> print(#str) -- 4
> print(str:len()) -- 4
> print(vim.fn.strlen(str)) -- 4
> print(#{str:byte(1, -1)}) -- 4
> ```
> 
> The same can be said for other multi-byte characters, like `€` (3) and `ñ` (2).
> 
> Why not just use `#`? Should be able to ditch `util.get_utf8_length` and then do:
> 
> ```diff
> diff --git a/lua/supermaven-nvim/binary/binary_handler.lua b/lua/supermaven-nvim/binary/binary_handler.lua
> index 7a83f42..d7a0fe7 100644
> --- a/lua/supermaven-nvim/binary/binary_handler.lua
> +++ b/lua/supermaven-nvim/binary/binary_handler.lua
> @@ -71,7 +71,7 @@ function BinaryLifecycle:on_update(buffer, file_name, event_type)
>    local cursor = api.nvim_win_get_cursor(0)
>    if cursor ~= nil then
>      local prefix = self:save_state_id(buffer, cursor, file_name)
> -    local offset = u.get_utf8_length(prefix)
> +    local offset = #prefix
>      updates[#updates + 1] = {
>        kind = "cursor_update",
>        path = file_name,
> ```

Correct, just verified it. I think `#str` and `str:len()` are the same thing, and `#str` is much faster than `strlen()`:
image

**Comment by Hashiraee** [comment]
I changed `vim.fn.strlen(str)` to `#str` as that is indeed much faster (thanks @b0o and @3rd). As for getting rid of the util function itself, I leave that up to @victorw-xyz.

With this everything should be taken care of now. I think it can be merged now @super-jacob?

**Comment by WilliamWelsh** [comment]
Thanks for fixing this, this issue has made it completely unusable for me in the meantime

**Pull request merged**
